Dynamic builds on Windows don't work

Dynamic builds on Windows don't work. Here is the error when trying to build with the windows-x64 architecture on base 3.14.12.5 and ADCore 2-5.  This was after already fixing problems in configure/RELEASE and pcoApp/src/Makefile:

```
make -C O.windows-x64 -f ../Makefile TOP=../../.. \
    T_A=windows-x64 install
make[3]: Entering directory 'P:/epics/epics/devel/areaDetector-2-5/ADPCO/pcoApp/src/O.windows-x64'
link /nologo /subsystem:windows /dll /LTCG /incremental:no /opt:ref /release   /MACHINE:X64      /implib:pco.lib /out:pco.dll          PCOSerial.obj pco.obj drvPcoEpics.obj drvpcoE
dgePluginEpics.obj pcoEdgePlugin.obj      P:\\epics\\epics\\devel\\areaDetector-2-5\\ADCameraLink\\lib\\windows-x64\\cameralink.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADCo
re\\lib\\windows-x64\\NDPlugin.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADCore\\lib\\windows-x64\\ADBase.lib  P:\\epics\\epics\\devel\\asyn-4-30\\lib\\windows-x64\\asyn.lib
 P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSupport\\lib\\windows-x64\\netCDF.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\lib\\windows-x64\\tiff.lib  P:\\epics\\epics\\devel\
\areaDetector-2-5\\lib\\windows-x64\\jpeg.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSupport\\lib\\windows-x64\\xml2.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSuppor
t\\lib\\windows-x64\\NeXus.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSupport\\lib\\windows-x64\\hdf5.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSupport\\lib\\windows
-x64\\szip.lib  P:\\epics\\epics\\devel\\areaDetector-2-5\\ADSupport\\lib\\windows-x64\\zlib.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\recIoc.lib  P:\\epics\\epics\\
base-3.14.12.5\\lib\\windows-x64\\softDevIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\miscIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\rsrvIoc.lib
P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\dbtoolsIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\asIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x6
4\\dbIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\registryIoc.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\dbStaticIoc.lib  P:\\epics\\epics\\base-3.14.
12.5\\lib\\windows-x64\\ca.lib  P:\\epics\\epics\\base-3.14.12.5\\lib\\windows-x64\\Com.lib gdi32.lib oleaut32.lib
   Creating library pco.lib and object pco.exp
PCOSerial.obj : error LNK2001: unresolved external symbol pasynOctetSyncIO
pco.obj : error LNK2001: unresolved external symbol pasynTrace
pco.dll : fatal error LNK1120: 2 unresolved externals
P:/epics/epics/base-3.14.12.5/configure/RULES_BUILD:300: recipe for target 'pco.dll' failed
make[3]: *** [pco.dll] Error 1120
make[3]: Leaving directory 'P:/epics/epics/devel/areaDetector-2-5/ADPCO/pcoApp/src/O.windows-x64'
P:/epics/epics/base-3.14.12.5/configure/RULES_ARCHS:64: recipe for target 'install.windows-x64' failed
make[2]: *** [install.windows-x64] Error 2
make[2]: Leaving directory 'P:/epics/epics/devel/areaDetector-2-5/ADPCO/pcoApp/src'
P:/epics/epics/base-3.14.12.5/configure/RULES_DIRS:87: recipe for target 'src.install' failed
make[1]: *** [src.install] Error 2
make[1]: Leaving directory 'P:/epics/epics/devel/areaDetector-2-5/ADPCO/pcoApp'
P:/epics/epics/base-3.14.12.5/configure/RULES_DIRS:87: recipe for target 'pcoApp.install' failed
make: *** [pcoApp.install] Error 2
```

There are 3 basic problems:
- Include files are in the wrong order with respect to epicsExport.h. That file must be the last file included before the include files that define functions that will be located in this DLL.
- Include files should really only include other include files that are needed for the definitions in that include file. Other include files should be included in the .cpp files, not in the .h files (e.g. , EPICS include files not defining items in this .h file, etc.). This is generally good style, and is very helpful when trying to get dynamic builds on Windows to work.
- Missing LIB_LIBS files. When building dynamically all external libraries must be defined when the library is built, using LIB_LIBS. This is not needed when building statically, or on non-Windows architectures, so it is frequently overlooked.

--- a/configure/RELEASE
+++ b/configure/RELEASE
@@ -7,3 +7,5 @@
 -include $(TOP)/../configure/RELEASE_LIBS.local
 -include $(TOP)/RELEASE.local
 -include $(TOP)/configure/RELEASE.local
+ADCAMERALINK=$(AREA_DETECTOR)/ADCameraLink
+

--- a/iocs/pcoIOC/configure/RELEASE
+++ b/iocs/pcoIOC/configure/RELEASE
@@ -9,3 +9,5 @@ ADPCO=$(TOP)/../..
 -include $(TOP)/../../../configure/RELEASE_PRODS.local
 -include $(TOP)/RELEASE.local
 -include $(TOP)/configure/RELEASE.local
+ADCAMERALINK=$(AREA_DETECTOR)/ADCameraLink
+

--- a/iocs/pcoIOC/pcoApp/src/Makefile
+++ b/iocs/pcoIOC/pcoApp/src/Makefile
@@ -9,8 +9,10 @@ include $(TOP)/configure/CONFIG
 
 
 PROD_NAME = pcoApp
-PROD_IOC_WIN32 += $(PROD_NAME)
-PROD_IOC_Linux += $(PROD_NAME)
+ifeq (windows-x64, $(findstring windows-x64, $(T_A)))
+  PROD_IOC_WIN32 += $(PROD_NAME)
+endif
+#PROD_IOC_Linux += $(PROD_NAME)
 
 
 # <name>.dbd will be created from <name>Include.dbd
@@ -18,13 +20,9 @@ DBD += $(PROD_NAME).dbd
 $(PROD_NAME)_DBD += PCOSupport.dbd
 $(PROD_NAME)_DBD  += camLinkSerialSupport.dbd
 $(PROD_NAME)_DBD  += pcoEdgePluginSupport.dbd
-$(PROD_NAME)_DBD  += caputRecorder.dbd
-
-
 
 # <name>_registerRecordDeviceDriver.cpp will be created from <name>.dbd
 PROD_SRCS_DEFAULT += $(PROD_NAME)_registerRecordDeviceDriver.cpp $(PROD_NAME)Main.cpp
-
 
 # Add locally compiled object code
 PROD_LIBS += pco clsersis display_lib fglib5 gbelib iolibrt 

--- a/iocs/pcoIOC/pcoApp/src/Makefile
+++ b/iocs/pcoIOC/pcoApp/src/Makefile
@@ -20,9 +20,9 @@ $(PROD_NAME)_DBD  += pcoEdgePluginSupport.dbd
 PROD_SRCS_DEFAULT += $(PROD_NAME)_registerRecordDeviceDriver.cpp $(PROD_NAME)Main.cpp
 
 # Add locally compiled object code
-PROD_LIBS += pco clsersis display_lib fglib5 gbelib iolibrt 
-
+PROD_LIBS += pco  
 PROD_LIBS += cameralink
+PROD_LIBS += clsersis display_lib fglib5 gbelib iolibrt
 
 include $(ADCORE)/ADApp/commonDriverMakefile
 

--- a/iocs/pcoIOC/pcoApp/src/Makefile
+++ b/iocs/pcoIOC/pcoApp/src/Makefile
@@ -3,11 +3,6 @@ include $(TOP)/configure/CONFIG
 #----------------------------------------
 #  ADD MACRO DEFINITIONS AFTER THIS LINE
 
-
-
-
-
-
 PROD_NAME = pcoApp
 ifeq (windows-x64, $(findstring windows-x64, $(T_A)))
   PROD_IOC_WIN32 += $(PROD_NAME)
@@ -27,11 +22,7 @@ PROD_SRCS_DEFAULT += $(PROD_NAME)_registerRecordDeviceDriver.cpp $(PROD_NAME)Mai
 # Add locally compiled object code
 PROD_LIBS += pco clsersis display_lib fglib5 gbelib iolibrt 
 
-PROD_LIBS += cameralink caputRecorder
-
-#!!$(foreach v, $(.VARIABLES), $(info $(v) = $($(v))))
-
-
+PROD_LIBS += cameralink
 
 include $(ADCORE)/ADApp/commonDriverMakefile
 

--- a/pcoApp/src/Makefile
+++ b/pcoApp/src/Makefile
@@ -13,8 +13,9 @@ include $(TOP)/configure/CONFIG
 INC += pco.h
 
 
-LIBRARY_IOC = pco
-
+ifeq (windows-x64, $(findstring windows-x64, $(T_A)))
+  LIBRARY_IOC = pco
+endif
 
 LIB_SRCS    += PCOSerial.cpp
 LIB_SRCS    += pco.cpp

--- a/pcoApp/src/Makefile
+++ b/pcoApp/src/Makefile
@@ -1,7 +1,6 @@
 TOP=../..
 include $(TOP)/configure/CONFIG
 
-
 #----------------------------------------
 #  
 #
@@ -12,20 +11,16 @@ include $(TOP)/configure/CONFIG
 
 INC += pco.h
 
-
 ifeq (windows-x64, $(findstring windows-x64, $(T_A)))
   LIBRARY_IOC = pco
 endif
 
 LIB_SRCS    += PCOSerial.cpp
 LIB_SRCS    += pco.cpp
-LIB_SRCS    +=  drvPcoEpics.c
+LIB_SRCS    += drvPcoEpics.c
 
-
-LIB_SRCS +=drvpcoEdgePluginEpics.c
-LIB_SRCS +=pcoEdgePlugin.cpp
-
-
+LIB_SRCS += drvpcoEdgePluginEpics.c
+LIB_SRCS += pcoEdgePlugin.cpp
 
 #one or the other
 #USR_CXXFLAGS += /D USE_SAP
@@ -39,44 +34,20 @@ LIB_SRCS +=pcoEdgePlugin.cpp
 USR_CXXFLAGS_Linux += -D USEASYNSERIAL
 USR_CXXFLAGS_WIN32 += /D USEASYNSERIAL
 
-
-
 #define to use the log_file class in camLinkSrc2, else it uses a null sort of class that does notbhgin
 USR_CXXFLAGS_WIN32 += /D USELOGFILE
 USR_CXXFLAGS_Linux += -D USELOGFILE
 
-
-
-
-
-#when we build debug.. we need this line?
-#!! for debvug only
-PROD_LDFLAGS_WIN32 = /NODEFAULTLIB:LIBCMT.lib
-
-
-
-
-PROD_LDFLAGS_WIN32 += /NODEFAULTLIB:msvcrt.lib
-
-
-PROD_LDFLAGS_WIN32 += /VERBOSE:LIB
-
 # Note that the order in which .dbd files are added to a list variable
 # doesn't matter, but the order of libraries does.
-
-
-#DBD += caputRecorder.dbd
-
-#!!DBD += PCOApp.dbd
-#DBD += PCOAppInclude.dbd
 
 DBD += PCOSupport.dbd
 DBD += pcoEdgePluginSupport.dbd
 
-LIB_LIBS += cameralink caputRecorder
+LIB_LIBS += cameralink
+LIB_LIBS += NDPlugin
 
 include $(ADCORE)/ADApp/commonLibraryMakefile
-
 
 include $(TOP)/configure/RULES
 

--- a/pcoApp/src/PCOSerial.cpp
+++ b/pcoApp/src/PCOSerial.cpp
@@ -10,8 +10,16 @@
  
 //#include "stdafx.h"
 
-#include "pco.h"
 #include <stdio.h>
+#include <math.h>
+
+#include <asynOctetSyncIO.h>
+#include <asynShellCommands.h>
+
+#include "ccd_exception.h"
+
+#include <epicsExport.h>
+#include "pco.h"
 
 // define if building for silicon softwarwe grabber.
 //#ifdef USE_SISW

--- a/pcoApp/src/pco.cpp
+++ b/pcoApp/src/pco.cpp
@@ -8,6 +8,61 @@
  *@date 2014
  */
 
+#include <stddef.h>
+#include <stdlib.h>
+#include <math.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <ctype.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+
+#include <epicsTime.h>
+#include <epicsThread.h>
+#include <epicsEvent.h>
+#include <epicsTimer.h>
+#include <epicsMutex.h>
+#include <epicsStdlib.h>
+#include <epicsString.h>
+#include <epicsStdio.h>
+#include <epicsMutex.h>
+#include <cantProceed.h>
+#include <iocsh.h>
+
+#include <asynOctetSyncIO.h>
+
+#ifdef USEASYNSERIAL
+#include <asynShellCommands.h>
+#endif
+
+#include "ADDriver.h"
+#define _ADIOC_
+#include "grabberInterface.h"
+
+#include "comportInterface.h"
+#include "logfile.h"
+#include "ccd_exception.h"
+#include "coreco.h"
+#include "stopWatch.h"
+#include "pco_structures.h"
+
+#ifdef USE_SISW
+#include "siswSerialPort.h"
+#endif
+
+#ifdef USE_SAP
+#include "cl_com_port2.h"
+#endif
+
+#include "ADCameralink.h"
+
+#ifndef _WIN32
+#define FALSE false
+#define TRUE true
+#endif
+
+#include <epicsExport.h>
 #include "pco.h"
 
 static const char *driverName = "pco";

--- a/pcoApp/src/pco.h
+++ b/pcoApp/src/pco.h
@@ -25,63 +25,11 @@
 #ifndef PCO_H_
 #define PCO_H_
 
-#include <stddef.h>
-#include <stdlib.h>
-#include <math.h>
-#include <stdio.h>
-#include <errno.h>
-#include <string.h>
-#include <ctype.h>
-#include <fcntl.h>
-#include <sys/stat.h>
-
-#include <epicsTime.h>
-#include <epicsThread.h>
 #include <epicsEvent.h>
-#include <epicsTimer.h>
-#include <epicsMutex.h>
-#include <epicsStdlib.h>
-#include <epicsString.h>
-#include <epicsStdio.h>
-#include <epicsMutex.h>
-#include <cantProceed.h>
-#include <iocsh.h>
-#include <epicsExport.h>
-
-#include <asynOctetSyncIO.h>
-
-#ifdef USEASYNSERIAL
-#include <asynShellCommands.h>
-#endif
-
-#include "ADDriver.h"
-#define _ADIOC_
-#include "grabberInterface.h"
-
-#include "comportInterface.h"
-#include "logfile.h"
-#include "ccd_exception.h"
-#include "coreco.h"
-#include "stopWatch.h"
-#include "pco_structures.h"
-
-#ifdef USE_SISW
-#include "siswSerialPort.h"
-#endif
-
-#ifdef USE_SAP
-#include "cl_com_port2.h"
-#endif
-
+#include <epicsTypes.h>
 #include "ADCameralink.h"
-#include <math.h>
-#include "time.h"
-
-
-#ifndef _WIN32
-#define FALSE false
-#define TRUE true
-#endif
+#include "pco_structures.h"
+#include "stopWatch.h"
 
 /** 
  *PCO Driver. Serial port functions. 
@@ -90,7 +38,7 @@
 //   putIntParam\((?<param>[a-z_]*).*
 // createParam\("$+{param}",asynInt32,$+{param}\)
 
-class pco : public ADCameralink {
+class epicsShareClass pco : public ADCameralink {
  public:
   pco(const char *portName, const char *pcoPort, int maxBuffers,
       size_t maxMemory, int priority, int stackSize);

--- a/pcoApp/src/pcoEdgePlugin.cpp
+++ b/pcoApp/src/pcoEdgePlugin.cpp
@@ -7,19 +7,22 @@
 #include <stdio.h>
 #include <string.h>
 #include <math.h>
-#include <netcdf.h>
+
+#include <string>
+using std::string;
+//#include <direct.h>
 
 #include <epicsStdio.h>
-#include <epicsExport.h>
 #include <iocsh.h>
 
 #include <epicsTime.h>
 #include <epicsThread.h>
 
 #include "NDPluginFile.h"
+
+#include <epicsExport.h>
 #include "pcoEdgePlugin.h"
 #include "drvpcoEdgePlugin.h"
-#include "asynNDArrayDriver.h"
 
 static const char *driverName = "pcoEdgePlugin";
 

--- a/pcoApp/src/pcoEdgePlugin.h
+++ b/pcoApp/src/pcoEdgePlugin.h
@@ -6,24 +6,13 @@
 #ifndef pcoEdgePlugin_H
 #define pcoEdgePlugin_H
 
-#include <stdlib.h>
-#include <string.h>
-#include <stdio.h>
-#include <errno.h>
-#include <sys/types.h>
-#include <sys/stat.h>
-
-#include <string>
-using std::string;
-//#include <direct.h>
-
-#include "NDPluginFile.h"
-
 /** Enums for plugin-specific parameters. There are currently no specific
  * parameters for this driver yet.
   * It uses the ADStdDriverParams and NDPluginDriver params. */
 
-class pcoEdgePlugin : public NDPluginFile {
+#include "NDPluginFile.h"
+
+class epicsShareClass pcoEdgePlugin : public NDPluginFile {
  public:
   pcoEdgePlugin(const char *portName, int queueSize, int blockingCallbacks,
                 const char *NDArrayPort, int NDArrayAddr, int priority,

--- a/pcoApp/src/pco_structures.h
+++ b/pcoApp/src/pco_structures.h
@@ -5,7 +5,13 @@
  *@author Timothy Madden
  *@date 2010
  */
- 
+
+#ifndef PCO_STRUCTURES_H_
+#define PCO_STRUCTURES_H_
+
+#ifdef _WIN32
+  #include <windows.h>
+#endif
 
 // ------------------------------------------------------------------------ //
 // -- Defines for Get Camera Type Command: -------------------------------- //
@@ -507,3 +513,5 @@ struct xPCO_Description {
 
   DWORD reserved[9];
 };
+
+#endif

